File: python/range-sum-of-bst.py
from typing import Optional

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
        

class Solution:
    def rangeSumBST(self, root: Optional[TreeNode], low: int, high: int) -> int:
        def dfs(node, low, high):
            if not node:
                return 0
            
            res = node.val if (low <= node.val <= high) else 0
            if node.val > low: res += dfs(node.left, low, high)
            if node.val < high: res += dfs(node.right, low, high)
            
            return res
        
        return dfs(root, low, high)

File: go/range_sum_of_bst.go
package leetcode

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func rangeSumBST(root *TreeNode, low int, high int) int {
	return dfs(root, low, high)
}

func dfs(node *TreeNode, low, high int) int {
	if node == nil {
		return 0
	}

	res := 0
	if node.Val >= low && node.Val <= high {
		res += node.Val
	}

	if node.Val > low {
		res += dfs(node.Left, low, high)
	}
	if node.Val < high {
		res += dfs(node.Right, low, high)
	}

	return res
}
